### PR TITLE
Realm

### PR DESCRIPTION
**Pull request by richard-pianka**

### CODE REVIEW

**Comment by richard-pianka** [comment]
- I'm not implementing games or ladder yet
- It turns out MCP does send a protocol byte, except it's 0x01, so I stuck it on a different socket; I chose port 6113 because it's what pvpgn uses
- Do you want the name "Olympus" as a setting?
- Sending a statstring for EID_USERUPDATE causes a rather esoteric rendering bug in Diablo II
  - We could make not sending it only a D2DV/D2XP thing if it's expected for other products
- Take a look at SID_ENTERCHAT lines 73-90, see comment
- The SID_LOGONREALMEX response has to send an IP address, I did a quick lookup but it's not exactly ideal...lmk if you're already getting this somewhere or you want a different solution (see NetworkUtilities.cs)
- I'm using a bunch of static but essentially random values for the MCP chunks (only actually relying on the cookie), not sure what you want to do here
- I had to essentially duplicate several classes: Message, Frame, one of the exceptions, the ServerSocket, etc.; any or all of these could be much more DRY with a refactor, lmk what you think (feels icky)
- There are a handful of places I want to decompose into functions, add helpers, etc. - planning to hit that in my next pass of refactors if that's cool
- I left a couple pieces of commented out code as reminders for places I need to treat when I add games and ladder
- There are like 4-5 areas where I found the bnetdocs documentation could be improved, I can submit these later and trim out my comments

File: src/Atlasd/Battlenet/ClientState.cs [hunk 1]
```diff
@@ -563,5 +565,27 @@ public void SocketIOCompleted_External(object sender, SocketAsyncEventArgs e)
...
+        // exists at this level because state split across two component objects
+        public byte[] GenerateDiabloIIStatstring()
+        {
+            var statstring = Product.ToByteArray(GameState.Product);
+
+            if (RealmState.ActiveCharacter != null)
+            {
+                var partial = new byte[][]
+                {
+                    "Olympus".ToBytes(),
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/ClientState.cs:578]: This should be made into a configurable. Perhaps you want to use this?

`var realmName = Settings.GetString(new string[] { "realm", "name" });`

File: src/Atlasd/Utilities/NetworkUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿using System.IO;
+using System.Net;
+
+namespace Atlasd.Utilities
+{
+    class NetworkUtilities
+    {
+        public static string GetPublicAddress()
+        {
+            WebRequest request = WebRequest.Create("https://api.ipify.org");
```
Comment (carlbennett) [inline src/Atlasd/Utilities/NetworkUtilities.cs:10]: The hardcoded url should be made into a configurable option.

Comment (carlbennett) [reply src/Atlasd/Utilities/NetworkUtilities.cs:10]: Around this line, there should also be a discriminator for IPv4, as the endpoint/url that a user may enter (or even override with their own DNS or hosts file) could contain an Dual-Stack (IPv4+IPv6 / A+AAAA) record.

Since Battle.net protocols only support IPv4, Atlas should make sure it only takes the IPv4 address, if both IPv4 and IPv6 are available from DNS, and it should error or give an empty/null/none result if there is only an IPv6 address with no IPv4 address from DNS or if DNS resolution fails.

File: src/Atlasd/Utilities/BytesHelper.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
+﻿using System;
+using System.Linq;
+using System.Net;
+using System.Text;
+
+namespace Atlasd.Helpers
+{
+    public static class BytesHelper
```
Comment (carlbennett) [inline src/Atlasd/Utilities/BytesHelper.cs:8]: This class seems fundamentally not necessary, as C# and .NET 3.1 already provide for many of the bit and byte helpers with `BitConverter` to start.

The methods below, GetBytes, AsString, ToBytes, that extend UInt32[], IPAddress, byte[], and strings, already exist elsewhere via the .NET namespaces.

Further, AsString() and ToBytes() assumes a UTF8 encoding and does not contain a parameter to set the encoding, not all of Battle.net supported UTF8 and there are components of the protocol that in fact used Latin1 (ISO-8859-1) encoding, despite Atlas's incorrect emulation only using UTF8. Any improvement to the general helpfulness of converting strings to byte arrays and byte arrays to strings, should include an encoding parameter at minimum, though again this entire class seems not necessary.

File: src/Atlasd/Daemon/Logging.cs [hunk 4]
```diff
@@ -85,6 +85,7 @@ public static void WriteLine(LogLevel level, LogType type, string buffer)
             if (level > CurrentLogLevel) return;
 
             Console.Out.WriteLine($"[{DateTime.Now}] [{LogLevelToString(level)}] [{LogTypeToString(type).Replace("_", "] [")}] {buffer}");
+            Console.Out.Flush();
```
Comment (carlbennett) [inline src/Atlasd/Daemon/Logging.cs:88]: I'm concerned about the performance of always calling Flush() when Debug log-level is set on a busy server.

File: src/Atlasd/Battlenet/RealmSocket.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        public RealmSocket(IPEndPoint localEndPoint)
+        {
+            IsListening = false;
+            Socket = null;
+            SetLocalEndPoint(localEndPoint);
+        }
+
+        public void Dispose() /* part of IDisposable */
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/RealmSocket.cs:30]: This needs to be refactored. It was copied from the original ServerSocket which is also wrong.

https://learn.microsoft.com/en-us/dotnet/api/system.idisposable?view=netcore-3.1

tl;dr - It needs to only Dispose during special conditions relating to the garbage collector, which this does not even consider.

File: src/Atlasd/Battlenet/Protocols/MCP/RealmState.cs [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+            //}
+
+            // Close the connection
+            try
+            {
+                if (Socket != null && Socket.Connected) Socket.Shutdown(SocketShutdown.Send);
+            }
+            catch (Exception ex)
+            {
+                if (!(ex is SocketException || ex is ObjectDisposedException)) throw;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/RealmState.cs:70]: I believe newer C# prefers not to re-throw exceptions and would prefer separate catch blocks for these two types of known exceptions. We could do that here since we're touching this part of the code.

File: src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_MOTD.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                            throw new RealmProtocolException(realmState.ClientState, $"{MessageName(Id)} must be sent from D2DV or D2XP");
+
+                        if (Buffer.Length != 0)
+                            throw new RealmProtocolException(realmState.ClientState, $"{MessageName(Id)} must be 0 bytes, got {Buffer.Length}");
+
+                        return new MCP_MOTD().Invoke(new MessageContext(realmState, MessageDirection.ServerToClient));
+                    }
+                case MessageDirection.ServerToClient:
+                    {
+                        var message = "Welcome to the Olympus realm server for Atlas!";
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_MOTD.cs:46]: This should be a configurable message from the json

File: src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_CREATEGAME.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+        public MCP_CREATEGAME(byte[] buffer)
+        {
+            Id = (byte)MessageIds.MCP_CREATEGAME;
+            Buffer = buffer;
+        }
+
+        public override bool Invoke(MessageContext context)
+        {
+            return false;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_CREATEGAME.cs:23]: Instead of returning false, this should have at least a logging line for debugging purposes that it wasn't even handled.

Returning false also I believe means to disconnect the connection from the client, being interpreted as they did something wrong and it wasn't generally successful.

File: src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_GAMEINFO.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+        public MCP_GAMEINFO(byte[] buffer)
+        {
+            Id = (byte)MessageIds.MCP_GAMEINFO;
+            Buffer = buffer;
+        }
+
+        public override bool Invoke(MessageContext context)
+        {
+            return false;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_GAMEINFO.cs:23]: Instead of returning false, this should have at least a logging line for debugging purposes that it wasn't even handled.

Returning false also I believe means to disconnect the connection from the client, being interpreted as they did something wrong and it wasn't generally successful.

File: src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_GAMELIST.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+        public MCP_GAMELIST(byte[] buffer)
+        {
+            Id = (byte)MessageIds.MCP_GAMELIST;
+            Buffer = buffer;
+        }
+
+        public override bool Invoke(MessageContext context)
+        {
+            return false;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_GAMELIST.cs:23]: Instead of returning false, this should have at least a logging line for debugging purposes that it wasn't even handled.

Returning false also I believe means to disconnect the connection from the client, being interpreted as they did something wrong and it wasn't generally successful.

File: src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_JOINGAME.cs [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+        public MCP_JOINGAME(byte[] buffer)
+        {
+            Id = (byte)MessageIds.MCP_JOINGAME;
+            Buffer = buffer;
+        }
+
+        public override bool Invoke(MessageContext context)
+        {
+            return false;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_JOINGAME.cs:23]: Instead of returning false, this should have at least a logging line for debugging purposes that it wasn't even handled.

Returning false also I believe means to disconnect the connection from the client, being interpreted as they did something wrong and it wasn't generally successful.

File: src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_NULL.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+        public MCP_NULL(byte[] buffer)
+        {
+            Id = (byte)MessageIds.MCP_NULL;
+            Buffer = buffer;
+        }
+
+        public override bool Invoke(MessageContext context)
+        {
+            return false;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_NULL.cs:23]: Instead of returning false, this should have at least a logging line for debugging purposes that it wasn't even handled.

Returning false also I believe means to disconnect the connection from the client, being interpreted as they did something wrong and it wasn't generally successful.

File: src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_REQUESTLADDERDATA.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+        public MCP_REQUESTLADDERDATA(byte[] buffer)
+        {
+            Id = (byte)MessageIds.MCP_REQUESTLADDERDATA;
+            Buffer = buffer;
+        }
+
+        public override bool Invoke(MessageContext context)
+        {
+            return false;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_REQUESTLADDERDATA.cs:23]: Instead of returning false, this should have at least a logging line for debugging purposes that it wasn't even handled.

Returning false also I believe means to disconnect the connection from the client, being interpreted as they did something wrong and it wasn't generally successful.

File: src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_CHARRANK.cs [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+        public MCP_CHARRANK(byte[] buffer)
+        {
+            Id = (byte)MessageIds.MCP_CHARRANK;
+            Buffer = buffer;
+        }
+
+        public override bool Invoke(MessageContext context)
+        {
+            return false;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_CHARRANK.cs:23]: Instead of returning false, this should have at least a logging line for debugging purposes that it wasn't even handled.

Returning false also I believe means to disconnect the connection from the client, being interpreted as they did something wrong and it wasn't generally successful.

File: src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_CHARLOGON.cs [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        public enum Statuses : UInt32
+        {
+            Success     = 0x00,
+            NotFound    = 0x46,
+            Failed      = 0x7A,
+            Expired     = 0x7B,
+        };
+
+        // set gamestate char name
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/Messages/MCP_CHARLOGON.cs:24]: ?

File: src/Atlasd/Battlenet/Protocols/MCP/MessageIds.cs [hunk 16]
```diff
@@ -13,7 +14,7 @@ enum Messages : byte
         MCP_REQUESTLADDERDATA = 0x11,
         MCP_MOTD = 0x12,
         MCP_CANCELGAMECREATE = 0x13,
-        MCP_CREATEQUEUE = 0x14,
+        MCP_CREATEQUEUE = 0x14, // not on bnetdocs
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/MCP/MessageIds.cs:17]: ???

This list of message ids was literally copied from bnetdocs.

Here's the webpage for it: https://bnetdocs.org/packet/308/mcp-createqueue

It's been there since 2010.

File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_QUERYREALMS2.cs [hunk 17]
```diff
@@ -42,12 +42,13 @@ public override bool Invoke(MessageContext context)
...
-                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
-
-                        Dictionary<byte[], byte[]> realms =
-                            context.Arguments == null || !context.Arguments.ContainsKey("realms") ?
-                            new Dictionary<byte[], byte[]>() :
-                            (Dictionary<byte[], byte[]>)context.Arguments["realms"];
+                        Dictionary<byte[], byte[]> realms = new Dictionary<byte[], byte[]>
+                        {
+                            {
+                                Encoding.UTF8.GetBytes("Olympus"),
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/Game/Messages/SID_QUERYREALMS2.cs:48]: Make this configurable from the server config json

File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_LOGONREALMEX.cs [hunk 18]
```diff
@@ -75,15 +77,84 @@ public override bool Invoke(MessageContext context)
...
+                                0x62653134,
+                                0x36313861,
+                                0x36353032,
+                                0x31353066,
+                                0x00000000,
+                                0x00000000
+                            }.GetBytes();
+
+                            w.Write(chunk2);
+                            w.WriteByteString(Encoding.UTF8.GetBytes(""));
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/Game/Messages/SID_LOGONREALMEX.cs:141]: `string.Empty` is preferred over `""`

File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_LOGONREALMEX.cs [hunk 19]
```diff
@@ -75,15 +77,84 @@ public override bool Invoke(MessageContext context)
...
+                            w.Write(ipAddress.GetBytes());
 
-                        w.Write((UInt32)clientToken);
-                        w.Write((UInt32)Statuses.RealmUnavailable); // Atlas does not have realm/MCP servers implemented yet.
+                            Settings.State.RootElement.TryGetProperty("battlenet", out var battlenetJson);
+                            battlenetJson.TryGetProperty("realm_listener", out var listenerJson);
+                            listenerJson.TryGetProperty("interface", out var interfaceJson);
+                            listenerJson.TryGetProperty("port", out var portJson);
+
+                            portJson.TryGetUInt16(out var port);
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/Game/Messages/SID_LOGONREALMEX.cs:115]: This should be refactored using the settings getter.

`ushort port = Settings.Get(new string[] { "battlenet", "interface", "port" }, default: 0);`

File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_LOGONREALMEX.cs [hunk 20]
```diff
@@ -75,15 +77,84 @@ public override bool Invoke(MessageContext context)
                          * (STRING)     Battle.net unique name (* as of D2 1.14d, this is empty)
                          */
 
-                        Buffer = new byte[8]; // MCP Cookie + MCP Status only; Atlas does not have realm/MCP servers implemented yet.
+                        if (((byte[])context.Arguments["realmTitle"]).AsString() == "Olympus")
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/Game/Messages/SID_LOGONREALMEX.cs:80]: Make it configurable

File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_ENTERCHAT.cs [hunk 21]
```diff
@@ -5,6 +5,7 @@
 using System.IO;
 using System.Linq;
 using System.Text;
+using Atlasd.Helpers;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/Game/Messages/SID_ENTERCHAT.cs:8]: Alphabetize the using list

File: src/Atlasd/Battlenet/Common.cs [hunk 22]
```diff
@@ -201,7 +209,35 @@ private static void InitializeListener()
...
-        
+
+        private static void InitializeRealmListener()
+        {
+            Settings.State.RootElement.TryGetProperty("battlenet", out var battlenetJson);
+            battlenetJson.TryGetProperty("realm_listener", out var listenerJson);
+            listenerJson.TryGetProperty("interface", out var interfaceJson);
+            listenerJson.TryGetProperty("port", out var portJson);
+
+            var listenerAddressStr = interfaceJson.GetString();
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Common.cs:220]: This should be refactored using the settings getter.

`string listenerAddressStr = Settings.Get(new string[] { "battlenet", "realm_listener", "interface", "port" }, default: string.Empty);`

File: src/Atlasd/Battlenet/Common.cs [hunk 23]
```diff
@@ -102,14 +107,17 @@ public static void Initialize()
             ActiveChannels = new ConcurrentDictionary<string, Channel>(StringComparer.OrdinalIgnoreCase);
             ActiveClans = new ConcurrentDictionary<byte[], Clan>();
             ActiveClientStates = new ConcurrentDictionary<Socket, ClientState>();
+            RealmClientStates = new ConcurrentDictionary<UInt32, ClientState>();
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Common.cs:110]: Alphabetize this

File: src/Atlasd/Battlenet/Common.cs [hunk 24]
```diff
@@ -42,13 +43,17 @@ public ShutdownEvent(string adminMessage, bool cancelled, DateTime eventDate, Ti
         public static ConcurrentDictionary<string, Channel> ActiveChannels;
         public static ConcurrentDictionary<byte[], Clan> ActiveClans;
         public static ConcurrentDictionary<Socket, ClientState> ActiveClientStates;
+        public static ConcurrentDictionary<UInt32, ClientState> RealmClientStates;
         public static List<GameAd> ActiveGameAds;
         public static ConcurrentDictionary<string, GameState> ActiveGameStates;
+        public static Realm Realm;
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Common.cs:49]: Alphabetize this by property name

**Review by carlbennett — CHANGES_REQUESTED** [review]
> I'm not implementing games or ladder yet

That's fine.

> It turns out MCP does send a protocol byte, except it's 0x01, so I stuck it on a different socket; I chose port 6113 because it's what pvpgn uses

That's fine.

> Do you want the name "Olympus" as a setting?

Yes.

> Sending a statstring for EID_USERUPDATE causes a rather esoteric rendering bug in Diablo II
> We could make not sending it only a D2DV/D2XP thing if it's expected for other products

It's expected that only users on the same product as the initiator will receive other users' EID_USERUPDATE messages, I believe. Spht may know more specifically on that. I'd be fine with D2DV/D2XP having special behavior if it's warranted, but only for statstring, as flag updates still need to occur for when a user gains channel operator after /designate or becomes channel speaker, etc.

> Take a look at SID_ENTERCHAT lines 73-90, see comment

Seems fine. Behavior is unchanged for now which I'm happy with. The comment seems like future work/todo.

> The SID_LOGONREALMEX response has to send an IP address, I did a quick lookup but it's not exactly ideal...lmk if you're already getting this somewhere or you want a different solution (see NetworkUtilities.cs)

I've commented already but the URL that it tries needs to be made configurable and it needs a discriminator for the type of IP address it uses to learn its public IP address.

> I'm using a bunch of static but essentially random values for the MCP chunks (only actually relying on the cookie), not sure what you want to do here

Not sure either. Seems like future work/todo. Static is fine at this stage where not everything is implemented.

> I had to essentially duplicate several classes: Message, Frame, one of the exceptions, the ServerSocket, etc.; any or all of these could be much more DRY with a refactor, lmk what you think (feels icky)

I've left comments throughout those files. Please address them. DRY would be better, it is icky right now.

> There are a handful of places I want to decompose into functions, add helpers, etc. - planning to hit that in my next pass of refactors if that's cool

I disagree with the premise of the ByteHelpers. I really want to look over each reason why you're adding one. The .NET standard library and C# language provide a lot already.

> I left a couple pieces of commented out code as reminders for places I need to treat when I add games and ladder

Seems fine.

> There are like 4-5 areas where I found the bnetdocs documentation could be improved, I can submit these later and trim out my comments

You asked to be made an editor on BNETDocs. I've already discussed this with you over Discord. I am not the sole decision maker on that. @nmbook and @Davnit also need to weigh their opinion as they are BNETDocs Staff, together all three of us make up a quorum. We've had editors make changes on the site without consultation and those changes ended up needing to be reversed, and I've made unilateral decisions in the past without asking others that ended up going bad for me. Fundamentally, I believe we'd rather moderate by having new changes made by the existing editors we have now. I'm open to opinions.
